governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,237 @@
+    }
+    
+    private Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    
+    private Map<Integer, ScopeCleanupAction> scopedCleanupActions = new LinkedHashMap<>();
+    private Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private ReferenceQueue<ScopeCleanupMarker> markerReferenceQueue = new ReferenceQueue<>();
+    private final ExecutorService reqQueueExecutor = Executors.newSingleThreadExecutor();
+    private final AtomicBoolean running;
+    
+    private final Provider<ScopeCleanupMarker> markerProvider = new Provider<ScopeCleanupMarker>() {
+        @Override
+        public ScopeCleanupMarker get() {
+            return new ScopeCleanupMarker();
+        }
+    };
+    
+    public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
+        this.scopeBindings = new HashMap<>(scopeBindings);
+        running = new AtomicBoolean(true);
+        reqQueueExecutor.submit(new Runnable() {
+            public void run() {
+                try {
+                    while (running.get()) {
+                            Reference<? extends ScopeCleanupMarker> ref = markerReferenceQueue.remove(1000);
+                            if (ref != null && ref instanceof ScopeCleanupAction) { 
+                                Integer markerKey = ((ScopeCleanupAction)ref).getId();
+                                ScopeCleanupAction cleanupAction;
+                                synchronized(scopedCleanupActions) {
+                                    cleanupAction = scopedCleanupActions.remove(markerKey);
+                                }
+                                if (cleanupAction != null) {
+                                    cleanupAction.call();
+                                }
+                            }
+                    }
+                } 
+                catch (InterruptedException e) {
+                }                
+            }
+        });
+    }
+    
+    public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
+        return binding.acceptScopingVisitor(new ManagedInstanceScopingVisitor(destroyableInstance, binding.getSource(), action));
+    }
+    
+    public void addScopeBindings(Map<Class<? extends Annotation>, Scope> bindings) {
+        scopeBindings.putAll(bindings);
+    }
+
+    /*
+     * compatibility-mode - scope is assumed to be eager singleton
+     */
+    public <T> boolean register(T destroyableInstance, Object context, Iterable<LifecycleAction> action) {
+        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
+    }
+    
+    /**
+     * final cleanup of managed instances if any
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        if (running.compareAndSet(true, false)) {
+            reqQueueExecutor.shutdown();
+            reqQueueExecutor.awaitTermination(2, TimeUnit.SECONDS);
```

Comment:
waiting is optional, to allow the single reqQueue thread an opportunity to
finish processing a single PreDestroy method invocation.  Worst case
scenario for the 2-second wait is that a single PreDestroy method
invocation could be interrupted, before the code below this block invokes
the rest of them.  Long / indefinite waits would potentially delay /
completely block this process.

On Wed, Jul 6, 2016 at 9:00 PM elandau <notifications@github.com> wrote:

> In
> governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java
> <https://github.com/Netflix/governator/pull/307#discussion_r69848666>:
>
> > +
> > +    /*
> > +     * compatibility-mode - scope is assumed to be eager singleton
> > +     */
> > +    public <T> boolean register(T destroyableInstance, Object context, Iterable<LifecycleAction> action) {
> > +        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
> > +    }
> > +
> > +    /**
> > +     * final cleanup of managed instances if any
> > +     */
> > +    @Override
> > +    public synchronized void close() throws Exception {
> > +        if (running.compareAndSet(true, false)) {
> > +            reqQueueExecutor.shutdown();
> > +            reqQueueExecutor.awaitTermination(2, TimeUnit.SECONDS);
>
> 2 seconds seems very arbitrary and likely to fail for any codebase that
> has a lengthy shutdown. We should either make this a configuration option,
> a much larger number (maybe 10 minutes) or just block indefinitely.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/governator/pull/307/files/5c39381496a87a3c2cac2fcca569b4f585e99859#r69848666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACIO-4oG6nzVMUcCsBp93OJMw3zcQDhWks5qTHnVgaJpZM4JGi7C>
> .
>